Fixes #38799 - Allow ForemanContext app_metadata to be extended from plugins

jeremylenz opened the pull request:
Foreman provides a React context, `ForemanContext`, that can be used anywhere in the React tree, including by plugins. This makes it easy to use data from the backend without the need for an API request. See https://github.com/theforeman/foreman/blob/develop/developer_docs/foreman-context.asciidoc

The actual data is set in `application_helper.rb` in Foreman. However, until now it was not possible for plugins to add to this data. With this PR, a procedure is introduced for plugins to register their app metadata during the plugin registration process.

For an example of use, see the Companion PR: https://github.com/theforeman/foreman_rh_cloud/pull/1116

jeremylenz committed:
Fixes #38799 - Allow ForemanContext app_metadata to be extended from plugins
Hide column display for non-relevant columns

Co-Authored-By: Chris Roberts <chrobert@redhat.com>

MariaAga commented on app/helpers/application_helper.rb:
```diff
@@ -407,7 +407,7 @@ def current_url_params(permitted: [])
     params.slice(*permitted.concat([:locale, :search, :per_page])).permit!
   end
 
-  def app_metadata
```
Need to change in `developer_docs/foreman-context.asciidoc`

MariaAga commented on webpack/assets/javascripts/react_app/components/ColumnSelector/helpers.js:
```diff
@@ -72,14 +74,21 @@ export const categoriesFromFrontendColumnData = ({
-    });
+    let shouldShowColumn = true;
+    if (typeof isRelevant === 'function') {
+      shouldShowColumn = isRelevant(contextData);
```
1. I dont see a usage of `isRelevant` in the rh_cloud PR?
2. could you add a comment of how it should be used?

jeremylenz replied:
Looks like those changes got lost @chris1984 

I will add them back manually

jeremylenz commented:
Addressed comments and updated both PRs.
Also DRYed up the column relevance checker function so that I can use it in 2 places
- column selector
- actual columns displayed in the table -- if a user has a saved table preference for a nonrelevant column, we still want to hide it. The next use of the column selector will remove the column anyway.

jeremylenz commented:
Added developer documentation.

MariaAga reviewed (commented):
The code looks ok, but I'm having troubles to set up the full env with rh_cloud and insights so I cant test it properly

jeremylenz commented:
@MariaAga could always just mock `ForemanRhCloud.with_iop_smart_proxy?` in foreman_rh_cloud `engine.rb` and that should be enough to manipulate the context value :)

chris1984 commented:
@MariaAga if you want me to test it I can, I worked with Jeremy on it. I can post the screenshots of the field not being there and the response in the api coming through. If you want to provide a green ack then :)

MariaAga commented:
@chris1984 That would be helpful, as my env doesnt start up so I cant mock rh_cloud :sweat_smile:

chris1984 reviewed (approved):
@MariaAga tested and it works correctly:
IOP box:

<img width="501" height="141" alt="Image" src="https://github.com/user-attachments/assets/a11a34f0-c425-44a7-b7a4-44862a569442" />

<img width="732" height="241" alt="Image" src="https://github.com/user-attachments/assets/738230c5-90ba-4993-9dc9-21f7bae0dce6" />

Non IOP box:

<img width="517" height="106" alt="Image" src="https://github.com/user-attachments/assets/76d66390-5126-4d57-a67c-516d48d31c12" />

<img width="719" height="226" alt="Image" src="https://github.com/user-attachments/assets/d655197b-b497-4a24-880b-b4b120c44326" />

MariaAga reviewed (approved):
Thanks everyone!
